scipy/sparse/linalg/_eigen/tests/test_svds.py

```diff
@@ -640,6 +646,26 @@ def test_svd_LM_zeros_matrix(self, shape, dtype):
         # Check that the singular values are zero.
         assert_array_equal(s, 0)
 
+    @pytest.mark.parametrize("shape", ((20, 20), (20, 21), (21, 20)))
```

Comment:
Can we actually test a sizable sparse array here which is going to be used for? This size is not nearly enough for anything to accumulate numerical errors to make the small SVDs to become problematic. Also 21 is not really different than 20 hence won't make any difference.

Reply:
@ilayn The only goal of this test is to demo a fail prior to the PR. It does fail and the failure is related to the previous poor approach rather than the error accumulation. 21 is just to make is non-square to check the code logic. 

I'll add a different more realistic test.

Reply:
Yes but at some point a sparse library needs sparse examples to make sure that they actually work. Otherwise dense algorithms will work anyways. This is just testing whether the syntax works. The claim of the PR is handling small SVs better but we don't know if that is the case for large sparse arrays.

Reply:
The PR doesn't change how sparse solvers perform. It only fixes currently poor postprocessing, which is agnostic to whether the problem is dense or sparse and large or small.

I agree that adding a new full featured unit test and/or example for small singular values may generally be a good idea and it might actually perform well, having this PR implemented. I would prefer however to add it in a separate new PR, to keep the current PR focused and finally merged.

Reply:
The accuracy of small singular values could be further improved if the suggested postprocessing is followed by a few iterations of lobpcg and another run of postprocessing, maybe in a higher precision compared to the first stage. Definitely ideas and considerations for a new PR...

Reply:
That's why I propose adding a decent sparse problem to test the post-processing. This is one of the key features of SV computation as the PR description claims but we are testing on tiny problems. That is certainly within the scope of this PR and needs to be tested. Otherwise this is just testing linear algebra tautologies since the size is too small. I have to disagree that this is not a distant problem.

Reply:
Okay, I will do it, thanks!

Reply:
We can use any example from sparse test matrix gallery for that. For example, take https://sparse.tamu.edu/Pajek/GD98_a

If you scroll through the images at the top right SVD structure is also provided and we can construct it both dense and sparse csc with 

```python
import numpy as np
from scipy.sparse import csc_array

rows = [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  1,  1,  1,  2,  2,  2,  2,
         4,  5,  5,  9,  9,  9,  9,  9,  9,  9,  9,  9,  9,  9, 10, 10, 10,
        10, 14, 14, 19, 21, 22, 22, 22, 23, 23, 23, 26, 26, 32, 34, 36]

cols = [ 1,  3,  4,  6,  8, 12, 17, 22, 29, 31,  0,  5,  9,  0,  7,  9, 13,
       37, 16, 27,  1,  4,  6, 11, 15, 17, 18, 22, 25, 28, 30,  0,  9, 16,
       26,  0,  9, 20,  0,  0,  9, 13, 16, 24, 27,  0,  9, 33, 35,  9]

Bdense = np.empty([38, 38], dtype=float)
Bdense[rows, cols] = 1.
Bsparse = csc_array((np.ones(50), (rows, cols)), shape=[38,38])
```

There are many examples in this suite that are designed specifically for these purposes and also we can download the big ones and import them at the testing times.

Reply:
Unit tests cannot use downloading. 

This PR needs singular values distributed in the 1e-10 to 1 range.

Reply:
Here is one example how we do it on some other test suite 

https://github.com/scipy/scipy/blob/main/scipy/linalg/tests/test_solvers.py#L113-L118=

We store the array in the test suite and at the test time we load for using it. Obviously, we can't add 1MB or so but a few kilobytes are not an issue. 

I am not sure why it is needed to be within `[1e-10, 1]`. It should handle small eigenvalues regardless no? Or am I missing the point?

Reply:
As anticipated, a new more practical test opens a different can of worms due to slow convergence and other issues with iterative solvers, which is totally unrelated to the given PR.

Reply:
> Yes but at some point a sparse library needs sparse examples to make sure that they actually work. Otherwise dense algorithms will work anyways. This is just testing whether the syntax works. The claim of the PR is handling small SVs better but we don't know if that is the case for large sparse arrays.

@ilayn I have added an example of this kind as a new unit tests and it passes after some massaging. Could you please have another look and if satisfied complete your review?